### PR TITLE
NOSTORY: merged both ecr public and private repository modules.

### PR DESCRIPTION
**Pull request by j0hnmwangi**
# Description
The purpose of this PR is to merge the public and private ECR modules into 1 module through creation of submodules and an example for each. This PR also aims to have the standard module structure and files added. 

## Features/Fixes/Patches/Changes list:
- ECR public sub-module
- minimum and complete examples
- `CHANGELOG.md` file
- `CODEOWNERS` file
- PR template file
- README footers
- ECR private module to submodule
- Terraform file to versions file
- Testing terraform binary to version `0.14.11`
- `pre-commit.yml` file
- `pre-commit-config.yaml` file

### PR Owners Checklist:
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you checked there aren't open [Pull Requests](../../../pulls) of upstream or parallel code that will cause conflicts?
* [x] Have you updated the `CHANGELOG.md`?
* [x] Have you updated the `README.md`(s)?
* [x] OWNER: Does your submission pass?
    * [x] Pre-commit (attach logs/print screen to this PR)
    * [x] Checkov/terrascan (attach logs/print screen to this PR)
    * [x] Terraform resources examples build/destroy successfully with Terraform 0.13.7 minimum?
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] First PR? Have you deleted `## How to use this template...` from the root `.README.md`?

### PR Reviewers Checklists?
* [ ] Are all your comments/changes resolved?
* [ ] Are you happy with the OWNER checklists and additional information provided?
* [ ] Does your review tests pass?
    * [ ] Terraform resources examples build/destroy successfully with Terraform 0.14.11 minimum?
```
       _               _              
   ___| |__   ___  ___| | _______   __
  / __| '_ \ / _ \/ __| |/ / _ \ \ / /
 | (__| | | |  __/ (__|   < (_) \ V / 
  \___|_| |_|\___|\___|_|\_\___/ \_/  
                                      
By bridgecrew.io | version: 2.0.1143 

terraform_plan scan results:

Passed checks: 7, Failed checks: 0, Skipped checks: 0

Check: CKV_AWS_227: "Ensure KMS key is enabled"
        PASSED for resource: aws_kms_key.main
        File: /tf.json:0-0
Check: CKV_AWS_7: "Ensure rotation for customer created CMKs is enabled"
        PASSED for resource: aws_kms_key.main
        File: /tf.json:0-0
        Guide: https://docs.bridgecrew.io/docs/logging_8
Check: CKV_AWS_33: "Ensure KMS key policy does not contain wildcard (*) principal"
        PASSED for resource: aws_kms_key.main
        File: /tf.json:0-0
        Guide: https://docs.bridgecrew.io/docs/ensure-kms-key-policy-does-not-contain-wildcard-principal
Check: CKV_AWS_51: "Ensure ECR Image Tags are immutable"
        PASSED for resource: aws_ecr_repository.this
        File: /tf.json:0-0
        Guide: https://docs.bridgecrew.io/docs/bc_aws_general_24
Check: CKV_AWS_136: "Ensure that ECR repositories are encrypted using KMS"
        PASSED for resource: aws_ecr_repository.this
        File: /tf.json:0-0
        Guide: https://docs.bridgecrew.io/docs/ensure-that-ecr-repositories-are-encrypted
Check: CKV_AWS_163: "Ensure ECR image scanning on push is enabled"
        PASSED for resource: aws_ecr_repository.this
        File: /tf.json:0-0
        Guide: https://docs.bridgecrew.io/docs/general_8
Check: CKV_AWS_32: "Ensure ECR policy is not set to public"
        PASSED for resource: aws_ecr_repository_policy.this
        File: /tf.json:0-0
        Guide: https://docs.bridgecrew.io/docs/public_1-ecr-repositories-not-public
```

### CODE REVIEW

**Review by mariomerco — DISMISSED** [review]
Looks good to me!

**Review by mariomerco — DISMISSED** [review]
Looks good to me!. There are two things I'll consider:

1. The KMS key for the private cross-region replication is not the same, which makes sense since the key I used is not multi-region. Either way, I don't know if worth taking a look at this so, if the key actually is multi-region, we could give the ability to reuse the same key (if ECR supports this feature, of course)

2. When destroying the Terraform stack, it doesn't delete the repository in the replica region, which could be a bit annoying. I understand the repository in the other region was really created by ECR and its cross-region replication feature, so Terraform doesn't know anything about it. But, maybe we can just let the note in the documentation mentioning that.

File: examples/complete/main.tf [hunk 1]
```diff
@@ -75,7 +78,9 @@ module "ecr" {
         {
             "Sid": "new policy",
             "Effect": "Allow",
-            "Principal": "*",
+            "Principal": {
+              "AWS" : "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:root"
+            },
```
Comment (patrickmukumbu) [inline examples/complete/main.tf]: In a previous PR, I was advised to have this line
`"AWS" : "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}"`
being
`"AWS" : "arn:${local.partition}:iam::${local.account_id}"`
and the locals being
locals {
partition = data.aws_partition.current.partition
account_id = data.aws_caller_identity.current.account_id
}

**Review by patrickmukumbu — CHANGES_REQUESTED** [review]

File: PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+### PR Owners Checklist:
+* [ ] Have you followed the guidelines in our Contributing document?
+* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
+* [ ] Have you checked there aren't open [Pull Requests](../../../pulls) of upstream or parallel code that will cause conflicts?
+* [ ] Have you updated the `CHANGELOG.md`?
+* [ ] Have you updated the `README.md`(s)?
+* [ ] OWNER: Does your submission pass?
+    * [ ] Pre-commit (attach logs/print screen to this PR)
+    * [ ] Checkov/terrascan (attach logs/print screen to this PR)
+    * [ ] Terraform resources examples build/destroy successfully with Terraform 0.13.7 minimum?
```
Comment (hugoalmeida264) [inline PULL_REQUEST_TEMPLATE.md:20]: Update the template, it is still pointing to 0.13.7

File: examples/complete/main.tf [hunk 3]
```diff
@@ -1,6 +1,8 @@
 
 locals {
-  name = "samplerepo"
+  name       = ["example-complete-repository", ]
+  partition  = data.aws_partition.current.partition
```
Comment (hugoalmeida264) [inline examples/complete/main.tf]: Locals should be on `locals.tf`

File: examples/complete/main.tf [hunk 4]
```diff
@@ -35,31 +37,24 @@ locals {
...
-  image_tag_mutability             = "IMMUTABLE"
-  create_replication_configuration = true
-  scan_type                        = "ENHANCED"
-  enable_pull_through_cache_rule   = true
-  ecr_repository_prefix            = "ecr-public"
-  upstream_registry_url            = "public.ecr.aws"
-  replica_region                   = "eu-west-2"
-  registry_id                      = data.aws_caller_identity.current.account_id
+module "private_ecr" {
+  for_each                  = toset(local.name)
```
Comment (hugoalmeida264) [inline examples/complete/main.tf]: Why is the name on a list? A string should be enough.

File: examples/complete/main.tf [hunk 5]
```diff
@@ -96,29 +93,4 @@ module "ecr" {
     ]
 }
```
Comment (hugoalmeida264) [inline examples/complete/main.tf]: I would like to see json code inside `jsonencode({})` over herodoc

File: examples/complete/main.tf [hunk 6]
```diff
@@ -1,6 +1,8 @@
 
 locals {
-  name = "samplerepo"
+  name       = ["example-complete-repository", ]
+  partition  = data.aws_partition.current.partition
+  account_id = data.aws_caller_identity.current.account_id
```
Comment (hugoalmeida264) [inline examples/complete/main.tf]: The untagged rule should be 1 day, by and large, you don't want untagged images, if they are there mist often than not it was by mistake or error and should be cleaned as soon as possible.

Comment (hugoalmeida264) [reply examples/complete/main.tf]: This is going to sound weird but `30` images to keep make more sense ;)

File: examples/complete/main.tf [hunk 7]
```diff
@@ -75,7 +70,9 @@ module "ecr" {
         {
```
Comment (hugoalmeida264) [inline examples/complete/main.tf]: Use `jsonencode({})`

File: main.tf [hunk 8]
```diff
@@ -1,105 +1,22 @@
...
-        }
-      }
-    }
-  }
+module "private" {
+  source                    = "./modules/private"
+  create_private_repository = var.create_private_repository
+  name                      = var.name
+  image_tag_mutability      = var.image_tag_mutability
+  scan_on_push              = var.scan_on_push
```
Comment (hugoalmeida264) [inline main.tf]: Should we remove this option since it is going to be deprecated?

File: main.tf [hunk 9]
```diff
@@ -1,105 +1,22 @@
...
+  encryption_configuration  = var.encryption_configuration
+  aws_ecr_lifecycle_policy  = var.aws_ecr_lifecycle_policy
+  aws_ecr_repository_policy = var.aws_ecr_repository_policy
+  environment               = var.environment
```
Comment (hugoalmeida264) [inline main.tf]: We should only have one tag and not mandatory ones

File: modules/private/main.tf [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# ECR Repository
+resource "aws_ecr_repository" "this" {
+  count                = var.create_private_repository ? 1 : 0
+  name                 = var.name
+  image_tag_mutability = var.image_tag_mutability
+  image_scanning_configuration {
+    scan_on_push = var.scan_on_push
```
Comment (hugoalmeida264) [inline modules/private/main.tf:7]: Should we leave this here.

File: modules/private/main.tf [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  dynamic "encryption_configuration" {
```
Comment (hugoalmeida264) [inline modules/private/main.tf:10]: Encryption should be enabled by default using at a min the `AWS Owned KMS key`

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]
See my comments.

Question, should we add a module for `aws_ecr_pull_through_cache_rule` here (Add it to the unreleased) or make it a separate module? It is still a repo but syncs from another repo.

**Comment by j0hnmwangi** [comment]
> See my comments.
> 
> Question, should we add a module for `aws_ecr_pull_through_cache_rule` here (Add it to the unreleased) or make it a separate module? It is still a repo but syncs from another repo.

I think it would be best to have it as part of the module that will have the other global settings, such as registry policies.

**Review by hugoalmeida264 — DISMISSED** [review]

**Review by patrickmukumbu — APPROVED** [review]

**Pull request merged**